### DISCUSSION TITLE
Key event to command palette seems to cancel pending workers?

### DISCUSSION THREAD

**Discussion by otherjason** [Textualize/textual#3613] [Q&A] [answered]
I have an app that uses several long-lived workers to asynchronously poll external services for data and update the UI. I found that closing the command palette, either via `ESC` or by selecting a command, seems to cancel any pending workers. Here's a small example:

```python
import asyncio
from textual import work
from textual.app import App, ComposeResult
from textual.widgets import RichLog

class Test(App):
    CSS = '''
    RichLog {
        height: 100%;
    }
    '''
    def compose(self) -> ComposeResult:
        yield RichLog(id = 'log')

        self.my_worker()

    @work(exclusive = False)
    async def my_worker(self):
        counter = 1

        log = self.query_one('#log')
        while True:
            log.write(f'test {counter}')
            counter += 1
            await asyncio.sleep(1)

if __name__ == "__main__":
    Test().run()
```

When running the above, it will print a new line to the `RichLog` every second. If you open the command palette, then the worker stays running. But it seems that after the palette comes up, as soon as I press a key, the worker stops running.

Is this behavior intended, or is this a bug?

**Comment by davep** [comment]
> Is this behavior intended, or is this a bug?

Absolutely isn't intended behaviour.

**Reply by davep** [reply]
My suspicion would be the use of `self.workers.cancel_all()` -- I think I thought that `self.workers` was all the workers on the current screen/widget. Looks like that might need double-checking.

**Comment by davep** [comment]
Raised as #3615.

**Comment by davep** [comment]
A fix is in the works: #3618

**Comment by davep** [comment, marked as answer]
Fixed in [0.41.0](https://github.com/Textualize/textual/releases/tag/v0.41.0).
